### PR TITLE
Handle keyboard and mouse events: move desktop preselection, close application and more

### LINKED ISSUE #9: Handle keyboards events: move highlight and execute functions
The window has to draw focus on start and receive keyboard events.
* `<Esc>` to close window
* Arrow keys and hjkl to move preselection of desktop
* `Enter` to open the desktop (execute dummy code for now)

### PR DESCRIPTION
**Pull request by TheHussky**
There are some keyboard & mouse events.
Feel free to test, try and add more.

### CODE REVIEW

File: src/dexpo.cpp [hunk 1]
```diff
@@ -61,39 +61,107 @@ main ()
...
+          {
+            xcb_motion_notify_event_t *mn
+                = reinterpret_cast<xcb_motion_notify_event_t *> (event);
+            int det = (mn->event_y - dexpo_padding)
+                      / v[0].height; // haha watch me name it like this again
+            highlighted = det;
+            w.highlight_window (0, dexpo_bgcolor);
+            w.highlight_window (1, dexpo_bgcolor);
+            w.highlight_window (2, dexpo_bgcolor);
+            w.highlight_window (highlighted, dexpo_hlcolor);
```
Comment (mmskv) [inline src/dexpo.cpp]: Create a member function that will efficiently redraw screen. Args: desktop number to highlight and action (highlight). So multiple hardcoded w.highlight_window won't be called everywhere

Also add a config variable for inactive desktops border color: for example highlight selected desktop in red and other desktops in grey

File: src/dexpo.cpp [hunk 2]
```diff
@@ -61,39 +61,107 @@ main ()
...
+              }
+            w.highlight_window (0, dexpo_hlcolor);
+            break;
+          }
+        case XCB_KEY_PRESS:
+          {
+            xcb_key_press_event_t *kp
+                = reinterpret_cast<xcb_key_press_event_t *> (event);
+            ;
+            if (kp->detail == 114) // right arrow
```
Comment (mmskv) [inline src/dexpo.cpp]: Add `hjkl` keybinds together with arrows.

Make both `Up arrow`, `Right arrow`, `l`, `k` preselect next desktop and
`Down arrow`, `Left arrow`, `h`, `j` preselect previous desktop
Add an `Enter` key handler (blank for now)

File: src/dexpo.cpp [hunk 3]
```diff
@@ -61,39 +61,107 @@ main ()
...
+                auto act_height = dexpo_padding;
+                for (const auto &dexpo_pixmap : v)
+                  {
+                    xcb_copy_area (window::c_, dexpo_pixmap.id, w.id,
+                                   desktop_pixmap::gc_, 0, 0, dexpo_padding,
+                                   act_height, dexpo_pixmap.width,
+                                   dexpo_pixmap.height);
+                    act_height += dexpo_pixmap.height;
+                    act_height += dexpo_padding;
+                  };
```
Comment (mmskv) [inline src/dexpo.cpp]: Do you think this should be reformatted and moved somewhere?
Looks like spaghetti

File: src/dexpo.cpp [hunk 4]
```diff
@@ -61,39 +61,107 @@ main ()
   window w (dexpo_x, dexpo_y, conf_width, conf_height);
   w.pixmaps = v;
   // Mapping pixmap onto window
-  while (1)
+  int highlighted = 0;
```
Comment (mmskv) [inline src/dexpo.cpp]: make this a `window` class attribute

File: src/dexpo.cpp [hunk 5]
```diff
@@ -61,39 +61,107 @@ main ()
...
+  xcb_generic_event_t *event;
```
Comment (mmskv) [inline src/dexpo.cpp:64]: Make smart

File: src/dexpo.cpp [hunk 6]
```diff
@@ -61,39 +61,107 @@ main ()
...
-              act_height += dexpo_padding;
-            };
-        }
-      w.highlight_window (1, dexpo_hlcolor);
-      /* We flush the request */
-      xcb_flush (window::c_);
+      xcb_flush (w.c_);
     }
-}
+}
```
Comment (mmskv) [inline src/dexpo.cpp:191]: Use clang-format

File: src/dexpo.cpp [hunk 7]
```diff
@@ -61,39 +61,107 @@ main ()
...
+            w.highlight_window (1, dexpo_bgcolor);
+            w.highlight_window (2, dexpo_bgcolor);
+            w.highlight_window (highlighted, dexpo_hlcolor);
+            if (kp->detail == 9) // escape
+              {
+                exit (0);
+              }
+            break;
+          }
+        case XCB_MOTION_NOTIFY: // pointer motion within window
```
Comment (mmskv) [inline src/dexpo.cpp:109]: Add a mouse click handler (leave blank for now)

Comment (mmskv) [reply src/dexpo.cpp:109]: Perhaps store all of the window's pixmap relative coordinates in a separate member struct?
Will this simplify mouse tracking? You won't have to iterate over the vectors.

File: src/dexpo.cpp [hunk 8]
```diff
@@ -61,39 +61,107 @@ main ()
...
+          {
+            xcb_motion_notify_event_t *mn
+                = reinterpret_cast<xcb_motion_notify_event_t *> (event);
+            int det = (mn->event_y - dexpo_padding)
+                      / v[0].height; // haha watch me name it like this again
```
Comment (mmskv) [inline src/dexpo.cpp]: Their height may be non-const when width is set in the config

File: src/dexpo.cpp (hunk 7 above)
Comment (mmskv) [reply src/dexpo.cpp:109]: Don't preselect  desktop unless mouse is above the pixmap.
Currently desktop preselects even when the mouse pointer is near the pixmap

**Review by mmskv — CHANGES_REQUESTED** [review]
I think a major rename is necessary to make naming consistent.
Currently we have gazillion names to describe a pixmap (image, screenshot) that is a downsized desktop (virtual desktop, workspace, monitor) screenshot: dexpo pixmap, desktop pixmap etc.

This is not our top priority but something to think about.

File: src/dexpo.cpp (hunk 3 above)
Comment (mangalinor) [reply src/dexpo.cpp]: Shouldn't vector refresh in cycle, so screenshots'll be more or less actual?

File: src/dexpo.cpp [hunk 9]
```diff
@@ -61,39 +61,107 @@ main ()
...
-                             dexpo_pixmap.height);
-              act_width += dexpo_pixmap.width;
-              act_width += dexpo_padding;
-            };
+              if (highlighted == 0)
+                {
+                  highlighted = v.size () ;
+                }
+                  highlighted -= 1;
+            }
```
Comment (mmskv) [inline src/dexpo.cpp]: just use mod

**Comment by mmskv** [comment]
Add customizable window border

File: src/dexpo.cpp (hunk 2 above)
Comment (mmskv) [reply src/dexpo.cpp]: Do not hardcode it button `ids`
Better query ids from X11 by key names (query 'space' and get int) and **cache them**

Caching is crucial. Do it with hash table or something.

File: src/dexpo.cpp (hunk 9 above)
Comment (TheHussky) [reply src/dexpo.cpp]: are you sure -1 mod v.size() =v.size()-1?

Comment (mmskv) [reply src/dexpo.cpp]: desktop ids start with zero. just as vector's indexes

File: src/dexpo.cpp [hunk 10]
```diff
@@ -104,45 +103,45 @@ main ()
...
-            w.highlight_window (2, dexpo_bgcolor);
+            if (kp->detail == 113
+                or kp->detail == 111) // left arrow or down arrow
+              {
+                w.highlight_window (highlighted, dexpo_bgcolor);
+                if (highlighted == 0)
+                  {
+                    highlighted = v.size ();
+                  }
+                highlighted -= 1;
```
Comment (mmskv) [inline src/dexpo.cpp]: why don't you just use modular subtraction

**Review by mmskv — CHANGES_REQUESTED** [review]

**Review by mangalinor — APPROVED** [review]

**Review by mmskv — APPROVED** [review]

**Pull request merged**